File: problems/414/solution.py
class Solution:
    def thirdMax(self, nums: List[int]) -> int:
        if len(nums) < 3:
            return max(nums)
        
        first, second, third = -sys.maxsize - 1, -sys.maxsize - 1, -sys.maxsize - 1
        
        for n in set(nums):
            if n > first:
                third, second, first = second, first, n
            elif n < first and n > second:
                third, second = second, n
            elif n < second and n > third:
                third = n
        
        if third == -sys.maxsize - 1:
            return first
        
        return third

File: problems/414/solution.go
package main
const MaxInt = int(^uint(0) >> 1)
const MinInt = -MaxInt - 1

func thirdMax(nums []int) int {
    first, second, third := MinInt, MinInt, MinInt
    
    for _, n := range nums {
        if n > first {
            third, second, first = second, first, n
        } else if n < first && n > second {
            third, second = second, n
        } else if n < second && n > third {
            third = n
        }
    }
    
    if third == MinInt {
        return first
    }
    
    return third
}